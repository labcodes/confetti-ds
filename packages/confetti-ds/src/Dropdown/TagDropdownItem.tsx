import React, {
  SyntheticEvent,
  useEffect,
  useCallback,
  MutableRefObject,
} from "react";
import { isEmpty } from "lodash";

import Icon from "../Icon";
import { IconTypes, TagTypes } from "../constants";
import AbstractTag from "../Tags/AbstractTag";

export interface BaseTagDropdownItemProps {
  /** This is the Tag's text. */
  text: string;
  /** Disables the Tag. Won't be read by screen readers. */
  disabled?: boolean;
  /** Action to be executed when the Tag is clicked. */
  onClick?: (event: SyntheticEvent) => any;
  /** Type of the icon to be rendered. Won't render an icon if not passed to the component. Can't have both 'icon' and 'thumbSrc' at the same time. */
  icon?: IconTypes;
  /** Sets Tag's color. */
  color?: TagTypes;
  /** Skin of the rendered Tag. */
  skin?: "pale" | "vivid";
  /** Sets an outline style. */
  isOutline?: boolean;
  /** Option tab index */
  tabIndex?: number;
  /** This function is used on AbstractTag's componenentDidMount to set the current Ref */
  setRef?: (ref: MutableRefObject<HTMLButtonElement>) => any;
  /** This function is used to handle click or keydown interactions */
  onInteraction?: ({
    event,
    ref,
  }: {
    event: any;
    ref: MutableRefObject<boolean>;
  }) => any;
  /** This is the option's value */
  value?: string;
  /** This prop is a boolean to verify if the option is current selected  */
  isSelected?: boolean;
}

export type TagDropdownItemProps =
  | (BaseTagDropdownItemProps & {
      /** Source of the thumb to be rendered. Won't render a thumb if not passed to the component. Can't have both 'icon' and 'thumbSrc' at the same time. */
      thumbSrc?: undefined;
      /** Alt text of the thumb to be rendered. */
      thumbAlt?: undefined;
    })
  | (BaseTagDropdownItemProps & {
      /** Source of the thumb to be rendered. Won't render a thumb if not passed to the component. Can't have both 'icon' and 'thumbSrc' at the same time. */
      thumbSrc: string;
      /** Alt text of the thumb to be rendered. */
      thumbAlt: string;
    });

export default function TagDropdownItem({
  text,
  disabled = false,
  onClick = () => {},
  icon,
  color,
  skin = "pale",
  isOutline = false,
  tabIndex = 0,
  setRef = () => {},
  onInteraction = () => {},
  value,
  isSelected = false,
  thumbSrc,
  thumbAlt,
}: TagDropdownItemProps) {
  const renderThumb = () =>
    thumbSrc ? (
      <img className="lab-tag__thumb" src={thumbSrc} alt={thumbAlt} />
    ) : undefined;

  const renderIcon = () =>
    icon ? (
      <Icon type={icon} color="black-75" className="lab-tag--left-icon" />
    ) : undefined;

  // super constructor
  const checkThumbAndIcon = useCallback(() => {
    const errorMessage =
      "`TagDropdownItem` can't be initialized with both `thumb` and `icon` props.";
    if (!isEmpty(thumbSrc) && !isEmpty(icon)) {
      throw new Error(errorMessage);
    }
  }, [thumbSrc, icon]);

  // componentDidUpdate
  useEffect(() => {
    checkThumbAndIcon();
  }, [thumbSrc, icon]);

  return (
    <AbstractTag
      text={text}
      className={`lab-tag--togglable${icon ? ` lab-tag--has-left-icon` : ""}${
        thumbSrc ? ` lab-tag--has-thumb` : ""
      }`}
      isOutline={isOutline}
      skin={skin}
      color={color}
      renderPrefix={renderThumb() || renderIcon()}
      onClick={onClick}
      disabled={disabled}
      tabIndex={tabIndex}
      onInteraction={onInteraction}
      isDropdown
      setRef={setRef}
      value={value}
      isSelected={isSelected}
      role="option"
    />
  );
}
