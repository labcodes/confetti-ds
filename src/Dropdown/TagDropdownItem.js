import React, { useEffect, useCallback } from "react";
import PropTypes from "prop-types";
import { isEmpty } from "lodash";

import Icon from "../Icon";
import { ICON_TYPES, TAG_COLORS } from "../constants";
import AbstractTag from "../Tags/AbstractTag";

export default function TagDropdownItem({
  text,
  thumbSrc,
  icon,
  color,
  skin,
  isOutline,
  onClick,
  disabled,
  tabIndex,
  onInteraction,
  setRef,
  value,
  isSelected,
}) {
  const checkThumbAndIcon = (thumbSrc, icon) => {
    const errorMessage =
      "`TagDropdownItem` can't be initialized with both `thumb` and `icon` props.";
    if (!isEmpty(thumbSrc) && !isEmpty(icon)) {
      throw new Error(errorMessage);
    }
  };

  // super constructor
  const handleCheckThumbAndIcon = useCallback(() => {
    checkThumbAndIcon();
  }, [icon, thumbSrc]);

  // componentDidUpdate
  useEffect(() => {
    handleCheckThumbAndIcon();
  }, []);

  const renderThumb = ({ thumbSrc }) =>
    thumbSrc ? (
      <img className="lab-tag__thumb" src={thumbSrc} alt="" />
    ) : undefined;

  const renderIcon = ({ icon }) =>
    icon ? (
      <Icon
        type={icon}
        color="black-75"
        size="petit"
        className="lab-tag--left-icon"
      />
    ) : undefined;

  return (
    <AbstractTag
      text={text}
      className={`lab-tag--togglable${icon ? ` lab-tag--has-left-icon` : ""}${
        thumbSrc ? ` lab-tag--has-thumb` : ""
      }`}
      isOutline={isOutline}
      skin={skin}
      color={color}
      renderPrefix={renderIcon(icon) || renderThumb(thumbSrc)}
      onClick={onClick}
      disabled={disabled}
      tabIndex={tabIndex}
      onInteraction={onInteraction}
      isDropdown
      setRef={setRef}
      value={value}
      isSelected={isSelected}
      role="option"
    />
  );
}

TagDropdownItem.propTypes = {
  /** This is the Tag's text. */
  text: PropTypes.string.isRequired,
  /** Disables the Tag. Won't be read by screen readers. */
  disabled: PropTypes.bool,
  /** Action to be executed when the Tag is clicked. */
  onClick: PropTypes.func,
  /** Source of the thumb to be rendered. Won't render a thumb if not passed to the component. Can't have both 'icon' and 'thumbSrc' at the same time. */
  thumbSrc: PropTypes.string,
  /** Type of the icon to be rendered. Won't render an icon if not passed to the component. Can't have both 'icon' and 'thumbSrc' at the same time. */
  icon: PropTypes.oneOf(ICON_TYPES),
  /** Sets Tag's color. */
  color: PropTypes.oneOf(TAG_COLORS),
  /** Skin of the rendered Tag. */
  skin: PropTypes.oneOf(["pale", "vivid"]),
  /** Sets an outline style. */
  isOutline: PropTypes.bool,
  /** Option tab index */
  tabIndex: PropTypes.string,
  /** This function is used on AbstractTag's componenentDidMount to set the current Ref */
  setRef: PropTypes.func,
  /** This function is used to handle click or keydown interactions */
  onInteraction: PropTypes.func,
  /** This is the option's value */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  /** This prop is a boolean to verify if the option is current selected  */
  isSelected: PropTypes.bool,
};

TagDropdownItem.defaultProps = {
  thumbSrc: "",
  icon: undefined,
  isOutline: false,
  skin: "pale",
  color: undefined,
  disabled: false,
  onClick: () => {},
  setRef: () => {},
  onInteraction: () => {},
  tabIndex: "0",
  isSelected: false,
};
