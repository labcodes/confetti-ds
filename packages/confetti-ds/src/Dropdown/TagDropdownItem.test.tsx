import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";

import { waitFor } from "@babel/core/lib/gensync-utils/async";
import TagDropdownItem from "./TagDropdownItem";

describe("TagDropdownItem", () => {
  it("renders with base props", async () => {
    const renderedComponent = renderer
      .create(
        <TagDropdownItem text="Test render TagDropdownItem" value="option1" />
      )
      .toJSON();

    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected with outline as true", async () => {
    const renderedComponent = renderer
      .create(
        <TagDropdownItem
          text="Test render outline TagDropdownItem"
          isOutline
          value="option1"
        />
      )
      .toJSON();

    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(
      <TagDropdownItem
        text="Test outline TagDropdownItem"
        isOutline
        value="option1"
      />
    ).html();
    expect(wrapper).toContain("lab-tag--outline");
  });

  it("renders as expected with a vivid skin", async () => {
    const renderedComponent = renderer
      .create(
        <TagDropdownItem
          text="Test render vivid TagDropdownItem"
          skin="vivid"
          value="option1"
        />
      )
      .toJSON();

    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(
      <TagDropdownItem
        text="Test vivid TagDropdownItem"
        skin="vivid"
        value="option1"
      />
    ).html();
    expect(wrapper).toContain("lab-tag--vivid");
    expect(wrapper).toContain('id="option1"');
  });

  it("renders as expected with a purple color", async () => {
    const renderedComponent = renderer
      .create(
        <TagDropdownItem
          text="Test render purple TagDropdownItem"
          color="purple"
          value="option1"
        />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(
      <TagDropdownItem
        text="Test vivid TagDropdownItem"
        skin="vivid"
        value="option1"
      />
    ).html();
    expect(wrapper).toContain("lab-tag--vivid");
    expect(wrapper).toContain('id="option1"');
  });

  it("renders as expected with a magnifying-glass icon", async () => {
    const renderedComponent = renderer
      .create(
        <TagDropdownItem
          text="Test render TagDropdownItem with magnifying-glass icon"
          icon="magnifying-glass"
          value="option1"
        />
      )
      .toJSON();

    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(
      <TagDropdownItem
        text="Test TagDropdownItem with magnifying-glass icon"
        icon="magnifying-glass"
        value="option1"
      />
    ).html();
    expect(wrapper).toContain("lab-icon--magnifying-glass");
    expect(wrapper).toContain('id="option1"');
  });

  it("renders as expected with a thumb", async () => {
    const renderedComponent = renderer
      .create(
        <TagDropdownItem
          text="Test render TagDropdownItem with thumb"
          thumbSrc="fake-thumb"
          value="option1"
        />
      )
      .toJSON();

    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(
      <TagDropdownItem
        text="Test TagDropdownItem with thumb"
        thumbSrc="fake-thumb"
        value="option1"
      />
    ).html();
    expect(wrapper).toContain(`src="fake-thumb"`);
    expect(wrapper).toContain('id="option1"');
  });

  it("does not render if passing both `thumb` and `icon` props", async () => {
    // Adding the async method to wait for the promises from the both useCallback and useEffect to be resolved.
    waitFor(() => {
      expect(() => {
        shallow(
          <TagDropdownItem
            text="Test to not render TagDropdownItem with with thumb and icon"
            icon="magnifying-glass"
            thumbSrc="fake-thumb"
            value="option1"
          />
        );
      }).toThrow(
        "`TagDropdownItem` can't be initialized with both `thumb` and `icon` props."
      );
    });
  });
});
