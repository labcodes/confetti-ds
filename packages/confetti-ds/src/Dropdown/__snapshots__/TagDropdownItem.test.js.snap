// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagDropdownItem renders as expected with a magnifying-glass icon 1`] = `
<button
  aria-selected={false}
  className="lab-tag lab-tag--togglable lab-tag--has-left-icon lab-tag--pale"
  disabled={false}
  id="Test render TagDropdownItem with magnifying-glass icon"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="option"
  tabIndex="0"
  type="button"
  value=""
>
  <span
    className="lab-icon lab-icon--magnifying-glass lab-icon--black-75 lab-icon--petit lab-tag--left-icon"
  />
  Test render TagDropdownItem with magnifying-glass icon
</button>
`;

exports[`TagDropdownItem renders as expected with a purple color 1`] = `
<button
  aria-selected={false}
  className="lab-tag lab-tag--togglable lab-tag--purple-pale"
  disabled={false}
  id="Test render purple TagDropdownItem"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="option"
  tabIndex="0"
  type="button"
  value=""
>
  Test render purple TagDropdownItem
</button>
`;

exports[`TagDropdownItem renders as expected with a thumb 1`] = `
<button
  aria-selected={false}
  className="lab-tag lab-tag--togglable lab-tag--has-thumb lab-tag--pale"
  disabled={false}
  id="Test render TagDropdownItem with thumb"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="option"
  tabIndex="0"
  type="button"
  value=""
>
  <img
    alt=""
    className="lab-tag__thumb"
    src="fake-thumb"
  />
  Test render TagDropdownItem with thumb
</button>
`;

exports[`TagDropdownItem renders as expected with a vivid skin 1`] = `
<button
  aria-selected={false}
  className="lab-tag lab-tag--togglable lab-tag--vivid"
  disabled={false}
  id="Test render vivid TagDropdownItem"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="option"
  tabIndex="0"
  type="button"
  value=""
>
  Test render vivid TagDropdownItem
</button>
`;

exports[`TagDropdownItem renders as expected with outline as true 1`] = `
<button
  aria-selected={false}
  className="lab-tag lab-tag--togglable lab-tag--outline lab-tag--pale"
  disabled={false}
  id="Test render outline TagDropdownItem"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="option"
  tabIndex="0"
  type="button"
  value=""
>
  Test render outline TagDropdownItem
</button>
`;

exports[`TagDropdownItem renders with base props 1`] = `
<button
  aria-selected={false}
  className="lab-tag lab-tag--togglable lab-tag--pale"
  disabled={false}
  id="Test render TagDropdownItem"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="option"
  tabIndex="0"
  type="button"
  value=""
>
  Test render TagDropdownItem
</button>
`;
